sysfs/class_fibrechannel.go

```diff
@@ -0,0 +1,248 @@
+	SupportedClasses string               // /sys/class/fc_host/<Name>/supported_classes
+	SupportedSpeeds  string               // /sys/class/fc_host/<Name>/supported_speeds
+	Counters         FibreChannelCounters // /sys/class/fc_host/<Name>/statistics/*
+}
+
+type FibreChannelClass map[string]FibreChannelHost
+
+// FibreChannelClass parses everything in /sys/class/fc_host.
+func (fs FS) FibreChannelClass() (FibreChannelClass, error) {
+	path := fs.sys.Path(fibrechannelClassPath)
+
+	dirs, err := ioutil.ReadDir(path)
+	if err != nil {
+		return nil, err
+	}
+
+	fcc := make(FibreChannelClass, len(dirs))
+	for _, d := range dirs {
+		host, err := fs.parseFibreChannelHost(d.Name())
+		if err != nil {
+			return nil, err
+		}
+
+		fcc[host.Name] = *host
+	}
+
+	return fcc, nil
+}
+
+// Parse a single FC host
+func (fs FS) parseFibreChannelHost(name string) (*FibreChannelHost, error) {
+	path := fs.sys.Path(fibrechannelClassPath, name)
+	host := FibreChannelHost{Name: name}
+
+	for _, f := range [...]string{"speed", "port_state", "port_type", "node_name", "port_id", "port_name", "fabric_name", "dev_loss_tmo", "symbolic_name", "supported_classes", "supported_speeds"} {
+		name := filepath.Join(path, f)
+		value, err := util.SysReadFile(name)
+		if err != nil {
+			return nil, fmt.Errorf("failed to read file %q: %v", name, err)
+		}
+
+		switch f {
+		case "speed":
+			host.Speed = value
+		case "port_state":
+			host.PortState = value
+		case "port_type":
+			host.PortType = value
+		case "node_name":
+			if len(value) > 2 {
+				value = value[2:]
+			}
+			host.NodeName = value
+		case "port_id":
+			if len(value) > 2 {
+				value = value[2:]
+			}
+			host.PortID = value
+		case "port_name":
+			if len(value) > 2 {
+				value = value[2:]
+			}
+			host.PortName = value
+		case "fabric_name":
+			if len(value) > 2 {
+				value = value[2:]
+			}
+			host.FabricName = value
+		case "dev_loss_tmo":
+			host.DevLossTMO = value
+		case "supported_classes":
+			host.SupportedClasses = value
+		case "supported_speeds":
+			host.SupportedSpeeds = value
+		case "symbolic_name":
+			host.SymbolicName = value
+		}
+	}
+
+	counters, err := parseFibreChannelStatistics(path)
+	if err != nil {
+		return nil, err
+	}
+	host.Counters = *counters
+
+	return &host, nil
+}
+
+// parseFibreChannelStatistics parses metrics from a single FC host.
+func parseFibreChannelStatistics(hostPath string) (*FibreChannelCounters, error) {
+	var counters FibreChannelCounters
+
+	path := filepath.Join(hostPath, "statistics")
+	files, err := ioutil.ReadDir(path)
+	if err != nil {
+		return nil, err
+	}
+
+	for _, f := range files {
+		if !f.Mode().IsRegular() || f.Name() == "reset_statistics" {
+			continue
+		}
+
+		name := filepath.Join(path, f.Name())
+		value, err := util.SysReadFile(name)
+		if err != nil {
+			if os.IsNotExist(err) || err.Error() == "operation not supported" || err.Error() == "invalid argument" {
```

Comment:
Can we add a comment why these errors are okay?

Reply:
Added!

Reply:
THanks!